core/system.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "global_include.h"
+#include "mavlink_include.h"
+
+namespace dronecore {
+
+class MAVLinkSystem;
+class DroneCoreImpl;
+class PluginImplBase;
+
+/**
+ * @brief This class represents a system, made up of one or more components
+ * (e.g. autopilot, cameras, servos, gimbals, etc). Commonly System objects
+ * are used to interact with UAVs (including their components) and standalone cameras.
+ */
+class System
+{
+public:
+    explicit System(DroneCoreImpl &parent, uint8_t system_id, uint8_t comp_id);
+    ~System();
+
+    /**
+     * @brief Checks whether the system is an autopilot.
+     * @return true if its an autopilot, false otherwise.
+     */
+    bool is_autopilot() const;
+
+    /**
+     * @brief Checks whether the system is a standalone (Non-autopilot).
+     * @return true if stand alone, false otherwise.
+     */
+    bool is_standalone() const;
+
+    /**
+     * @brief Checks whether the system has camera with the given camera ID.
+     *
+     * A System may have several cameras with ID starting from 1.
+     * When called without passing camera ID, it checks for the camera ID 1.
+     * @param camera_id ID of the camera starting from 1 onwards.
+     * @return true if camera with the given camera ID is found, false otherwise.
+     */
+    bool has_camera(uint8_t camera_id = 1) const;
+
+    /**
+     * @brief Checks whether the system has gimbal.
+     * @return true if the system has gimbal, false otherwise.
+     */
+    bool has_gimbal() const;
+
+    friend class DroneCoreImpl;
```

Comment:
Why are the friends needed? Please add a comment explaining why.

Reply:
Oh, I had not seen that. I'm interesting in why we need friends at all. I thought they were made for operators, but that they were not good practice in other uses...

Reply:
Yeah good point.

- `friend class PluginImplBase`:
While we create plugins, we pass a reference to `System` to plugin constructors. But eventually, plugins need access to `MAVLinkSystem` instance. So, we keep `MAVLinkSystem` as a private member of class `System`.
By making `DroneCoreImpl` class a friend of class `System`, we allow to `DroneCoreImpl` class to request access to instance of `MAVLinkSystem` class.
- `friend class DroneCoreImpl`:
class `DroneCoreImpl` wants to perform some operations on class `System`, but we don't want those to expose to public. Please suggest for alternative if any.

Reply:
Usually we use used the pimpl scheme to solve exactly this. I guess we can leave it for now but might want to optimize this later.

Reply:
OK